### ISSUE TITLE
How to make package managers happy

### ISSUE THREAD

**Issue by seirl** [cmangos/issues#1961]
I'm working on making an Archlinux package for CMangos (and maybe later packages for other distributions). While writing my PKGBUILD, I noticed a lot of issues that I had to work around (some of which I just couldn't) in the installation process that made it hard to properly package CMangos according to proper POSIX/FHS standards.

The goal of this meta-issue is to track the different roadblocks I encountered, and to update the TODO-list as we make progress on this.

- [ ] **Config directory**: Right now, when installing in the /usr prefix (`-DCMAKE_INSTALL_PREFIX=/usr`), the config ends up in /usr/etc instead of /etc. Simply using SYSCONFDIR and an empty prefix instead of manually redefining a CONF_DIR variable like is done [here](https://github.com/cmangos/mangos-wotlk/blob/master/CMakeLists.txt#L143) should be enough in theory, as the [cmake docs](https://cmake.org/cmake/help/v3.4/module/GNUInstallDirs.html) state:

    > For `<dir>` equal to SYSCONFDIR or LOCALSTATEDIR, the `CMAKE_INSTALL_FULL_<dir>` is computed by prepending just / to the value of `CMAKE_INSTALL_<dir>` if it is not user-specified as an absolute path. For example, the `SYSCONFDIR` value etc becomes /etc.

- [ ] **Config installed as .conf.dist**: I get why it's useful to install the config as .conf.dist files, but for packages it's very bad practice to not have the config installed properly by default. Most package managers already have a way to handle conflicts to avoid overriding user config when installing (e.g `.new` files in Pacman). Right now I have to manually move the files like this after the package creation:

      mv "${pkgdir}/etc/realmd.conf.dist" "${pkgdir}/etc/realmd.conf"

  It would be good to have a configure flag that allows to directly install the files as .conf instead of .conf.dist

- [ ] **Tools installed in bin/tools**: AFAICT it's currently impossible to change the destination dir of the maps/vmaps/dbc/... extraction tools. The path is hardcoded in places like [this](https://github.com/cmangos/mangos-wotlk/blob/master/contrib/mmap/CMakeLists.txt#L106) to be `$prefix/bin/tools`. This is obviously doesn't make much sense when installing stuff system-wide, as users are not going to call "tools/something" on the command line. Plus, polluting the global PATH namespace with a random "tools" directory is a bad idea.

  I don't think it's necessary to have the tools in $PREFIX. A better place for these tools would probably be `/var/lib/mangos/tools/`, and users would just call/copy the tools installed by the package with this absolute path. Therefore we should make it possible to install the tools in this place.

- [ ] **Tools require write access on a WoW folder**: This is a pretty big one. Right now the workflow is to copy the extraction tools inside the WoW folder, to run them from this working directory and then to copy back the extracted files in some place. The thing is, this is a mess for UNIX permissions. Generally the user with which I'm going to run the extraction tools isn't the same as the owner of the WoW directory. Considering its size, copying it in my state directory isn't an option either, it would be a complete waste of space.

  Here, the best workflow would be to just have a way to specify a path to the WoW folder, and potentially also a path to the output folder that defaults to the current directory. e.g:

    /var/lib/mangos/tools/ExtractResources.sh a /opt/wow-3.3.5a /var/lib/mangos

  (where /opt/wow-3.3.5a is the game folder, and /var/lib/mangos is the output directory where we want to put maps/vmaps/dbc/...). Doing this actually requires two different issues to be fixed:

    - [ ] Add a way to configure the game directory and output directory for every single tool and in the `ExtractResources.sh` wrapper
    - [ ] Change the way the scripts call themselves together to not use a relative path compared to the *current working directory*, but to the *current script absolute path* (i.e `$( dirname $0 )/$script_name`)

- [ ] **run-mangosd is useless**: Most distributions use systemd nowadays, making scripts like run-mangosd obsolete. There should be a configure switch to not install this script.

- [ ] **Reproducible builds**: Right now, the use of the __FILE__ macro makes it so that some of the paths that were used at build time are included as strings in the binaries. This makes builds non-reproducible, can be a privacy issue and make package managers not happy, with errors like "==> WARNING: Package contains reference to $srcdir". I'm not sure what a proper fix would be, maybe the usage of __FILE__ could simply be removed. There's also apparently a -ffile-prefix-map that could be used to solve this problem:

      -ffile-prefix-map=old=new
      When compiling files residing in directory old, record any references to them in the result of the compilation as if the files resided in directory new instead.

**Comment by berserkingyadis** [comment]
:+1:

**Comment by cyberium** [comment]
Look promising. However i will directly warn you about this will never be merged if you dev it/test it only on archlinux.
 It should be designed in a way that at least ubuntu/debian/and maybe macosx? are supported too.
Ofc always start with a system that you know/like better.

Just wanted to point that out.

Cheers

**Comment by Muehe** [comment]
@cyberium The only thing Arch-specific here is the PKGBUILD file, but adding that would have no effect on the other distributions or Windows/Mac. So I don't really see a reason to not add it.

All the other stuff discussed here would be beneficial for all Linux distributions.

@seirl A cursory look over your files leads me to believe they are not independent of expansion, i.e. You can install Wotlk with it, but if you were to install Classic/TBC in a similar way afterwards it would overwrite the Wotlk install.

**Files mentioned in Discord:**
https://github.com/seirl/prologin-sadm-wow/blob/master/pkg/PKGBUILD
https://github.com/seirl/prologin-sadm-wow/blob/master/pkg/mangos-extract-resources

PKGBUILD depends on mariadb, what about people who do/want to use mysql?

Adding a new extraction script seems like overkill, I think we should edit the existing script (and the binaries if necessary) to accept location parameters.

**Comment by Muehe** [comment]
P.S.: @seirl FYI there are plans of making realmd/authserver its own repo that works across all versions, not sure how far along this plan is though. @cyberium maybe able to say more on this.

**Comment by cyberium** [comment]
Iam not aware of all distrib difference for sure so i'll have to rely on you for that.
Yes iam still waiting for news about it. Something should come in a way or another for a stand alone authserver.

**Comment by seirl** [comment]
@Muehe it's not my intention right now to merge the arch specific stuff into CMangos or whatever, this was just for my personal usage. I agree it has to be made aware of the expansion if we were to make it more largely avaible, but this is unrelated to this issue.

**Comment by jimmybrancaccio** [comment]
If this is not going into our repositories is there any reason we have issues for it on our tracker instead of your own?

**Comment by berserkingyadis** [comment]
I think he was referring to the arch specific stuff. The improvements to for example the extractors or cmake will be merged into the repo, no? @seirl

**Comment by seirl** [comment]
Yes of course, all of these are upstream problems that make it hard to package cmangos and that should be fixed in cmangos.

**Comment by esno** [comment]
hey, I started working to address such issues with package management for linux distros a few days ago and found this topic. ATM it's focused on gentoo but using GNUINSTALLDIR variables should be at least distro-agnostic - don't know how windows deals with that and if it breaks user workflows. would be great if someone could test it.
https://github.com/cmangos/mangos-tbc/compare/master...blizzlike-org:bugfix/cmake_install_path
https://github.com/cmangos/mangos-classic/compare/master...blizzlike-org:bugfix/cmake_install_path
@Muehe As for now I'm not sure if file conflicts between vanilla/tbc/wotlk should be handled in cmake or by renaming such files in post install scripts (pkg manager)

**Comment by seirl** [comment]
This looks like a great start! Thanks a lot for this.

**Comment by Muehe** [comment]
@esno I'm not sure either, but for many of the sub-issues this is blocking. I would just add post-fix to all the files involved ( `mangosd` → `mangosd-classic`, `mangosd.conf` → `mangosd-classic.conf`, etc.) but I would like to have opinions from @cyberium and @killerwife before I proceed.

**Comment by esno** [comment]
and thats where I need your input. I'm fine with both. for my gentoo packages I already prepend `-cmangos-<flavor>` to all potentially conflicting files but prepending that in cmake would also be fine. I'll walk the path you guys prefer.

**Comment by esno** [comment]
relates to https://github.com/cmangos/issues/issues/1873

**Comment by esno** [comment]
relates to https://github.com/cmangos/issues/issues/1166

**Comment by Zombie-Ryushu** [comment]
My family members are OpenSuse Packagers. We are working on a RPM Specfile universal for all RPM Distros using the OpenSuse Build Service. Any assistance.

**Comment by al3xc1985** [comment]
@seirl still valid your intentions?

**Comment by Muehe** [comment]
@al3xc1985 Yes this issue is still valid. The issue list in the OP (with one small edit I just made) is still valid. Anybody is welcome to progress the sub-issues with PRs or triage.
